idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/utils/IdeUtils.kt

```diff
@@ -1,16 +1,45 @@
 package arrow.meta.ide.dsl.utils
 
+import com.intellij.psi.PsiElement
+import com.intellij.psi.SyntaxTraverser
 import org.celtric.kotlin.html.BlockElement
 import org.celtric.kotlin.html.InlineElement
 import org.celtric.kotlin.html.code
 import org.celtric.kotlin.html.text
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtPsiFactory
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
 object IdeUtils {
   fun <A> isNotNull(a: A?): Boolean = a?.let { true } ?: false
 }
 
+/**
+ * traverse and filters starting from the root node [receiver] down to all it's children and applying [f]
+ */
+fun <A : PsiElement, B> PsiElement.traverseFilter(on: Class<A>, f: (A) -> B): List<B> =
+  SyntaxTraverser.psiTraverser(this).filter(on).map(f).toList()
+
+/**
+ * Modify [element] based on [f]
+ */
+inline fun <reified K : KtElement> KtPsiFactory.modify(element: K, noinline f: KtPsiFactory.(K) -> K?): K? =
+  f(this, element)?.run { element.replaceK(this) }
+
+/**
+ * reified PsiElement replacement
+ */
+inline fun <reified K : PsiElement> K.replaceK(to: K): K? =
```

Comment:
what does this do that astReplace or PsiElement replace doesn't already provide? Is it just the null check?

Reply:
No, I needed a reified version of that function in order to compose them.